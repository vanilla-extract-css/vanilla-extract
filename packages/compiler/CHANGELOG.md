# @vanilla-extract/compiler

## 0.1.3

### Patch Changes

- [#1583](https://github.com/vanilla-extract-css/vanilla-extract/pull/1583) [`6488e28`](https://github.com/vanilla-extract-css/vanilla-extract/commit/6488e28337106582231b418778c884f26e393219) Thanks [@askoufis](https://github.com/askoufis)! - Ignore Vite `base` config

  Fixes a bug where imported asset URLs would be transformed by the compiler when setting the `base` option in Vite, resulting in incorrect resolution and bundling of these assets during the build process.

- Updated dependencies [[`134117d`](https://github.com/vanilla-extract-css/vanilla-extract/commit/134117d32e25d92edc24c8c863d757294d7b5df8), [`c66be53`](https://github.com/vanilla-extract-css/vanilla-extract/commit/c66be53d600802b2922da1d6034e2a5ff3fbbcae)]:
  - @vanilla-extract/css@1.17.2
  - @vanilla-extract/integration@8.0.2

## 0.1.2

### Patch Changes

- Updated dependencies [[`965fd03`](https://github.com/vanilla-extract-css/vanilla-extract/commit/965fd03ff26dd324ec24734aa7700f1fe89bd483)]:
  - @vanilla-extract/integration@8.0.1

## 0.1.1

### Patch Changes

- [#1529](https://github.com/vanilla-extract-css/vanilla-extract/pull/1529) [`d5b6e70`](https://github.com/vanilla-extract-css/vanilla-extract/commit/d5b6e70f44a3d4f03e113fe78e0605b358e9c0d7) Thanks [@askoufis](https://github.com/askoufis)! - Update `vite-node` dependency

- [#1529](https://github.com/vanilla-extract-css/vanilla-extract/pull/1529) [`d5b6e70`](https://github.com/vanilla-extract-css/vanilla-extract/commit/d5b6e70f44a3d4f03e113fe78e0605b358e9c0d7) Thanks [@askoufis](https://github.com/askoufis)! - Include `^6.0.0` in `vite` dependency range

## 0.1.0

### Minor Changes

- [#1536](https://github.com/vanilla-extract-css/vanilla-extract/pull/1536) [`a8248be`](https://github.com/vanilla-extract-css/vanilla-extract/commit/a8248befac51aa51d771b9b22a46209b1fd1e3b3) Thanks [@askoufis](https://github.com/askoufis)! - Initial release

### Patch Changes

- Updated dependencies [[`5f66abb`](https://github.com/vanilla-extract-css/vanilla-extract/commit/5f66abbd607e76d491bbb7b9bfe9c64c882a53e8), [`a8248be`](https://github.com/vanilla-extract-css/vanilla-extract/commit/a8248befac51aa51d771b9b22a46209b1fd1e3b3), [`c432ff3`](https://github.com/vanilla-extract-css/vanilla-extract/commit/c432ff33a8aead2c94fa6a4fcc9fcb1d7990427d), [`ec0b024`](https://github.com/vanilla-extract-css/vanilla-extract/commit/ec0b024fd19c133c233445f9e860626d104f9d97), [`a8248be`](https://github.com/vanilla-extract-css/vanilla-extract/commit/a8248befac51aa51d771b9b22a46209b1fd1e3b3)]:
  - @vanilla-extract/integration@8.0.0
  - @vanilla-extract/css@1.17.1
