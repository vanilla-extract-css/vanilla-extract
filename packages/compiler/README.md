# @vanilla-extract/integration

This package is not intended for public consumption.