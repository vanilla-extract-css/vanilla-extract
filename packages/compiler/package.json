{
  "name": "@vanilla-extract/compiler",
  "version": "0.3.0",
  "description": "Zero-runtime Stylesheets-in-TypeScript",
  "main": "dist/vanilla-extract-compiler.cjs.js",
  "module": "dist/vanilla-extract-compiler.esm.js",
  "types": "dist/vanilla-extract-compiler.cjs.d.ts",
  "files": [
    "/dist"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/vanilla-extract-css/vanilla-extract.git",
    "directory": "packages/compiler"
  },
  "author": "SEEK",
  "license": "MIT",
  "dependencies": {
    "@vanilla-extract/css": "workspace:^",
    "@vanilla-extract/integration": "workspace:^",
    "vite": "^5.0.0 || ^6.0.0 || ^7.0.0",
    "vite-node": "^3.2.2"
  }
}
