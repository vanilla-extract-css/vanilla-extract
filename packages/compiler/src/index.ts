export {
  createCompiler,
  type Compiler,
  type CreateCompilerOptions,
} from './compiler';
