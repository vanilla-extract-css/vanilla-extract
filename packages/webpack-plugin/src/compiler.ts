import type { LoaderContext } from './types';
import createCompat from './compat';

// Should be "ExternalsItem" but webpack doesn't expose it
type Externals = any;

interface CompilationResult {
  source: string;
  fileDependencies: Array<string>;
  contextDependencies: Array<string>;
}

const getCompilerName = (resource: string) =>
  `vanilla-extract-compiler:${resource}`;

export class ChildCompiler {
  externals: Externals | undefined;

  constructor(externals: Externals) {
    this.externals = externals;
  }

  isChildCompiler(name: string | undefined) {
    return (
      typeof name === 'string' && name.startsWith('vanilla-extract-compiler')
    );
  }

  async getCompiledSource(loader: LoaderContext) {
    const { source, fileDependencies, contextDependencies } =
      await compileVanillaSource(loader, this.externals);

    // Set loader dependencies to dependencies of the child compiler
    fileDependencies.forEach((dep) => {
      loader.addDependency(dep);
    });
    contextDependencies.forEach((dep) => {
      loader.addContextDependency(dep);
    });

    return {
      source,
      dependencies: fileDependencies,
    };
  }
}

function getRootCompilation(loader: LoaderContext) {
  var compiler = loader._compiler;
  var compilation = loader._compilation;
  while (compiler.parentCompilation) {
    compilation = compiler.parentCompilation;
    compiler = compilation.compiler;
  }
  return compilation;
}

// Modified version of webpack's regex for detecting template strings.
// We only want to match un-escaped strings so we can escape them ourselves.
// https://github.com/webpack/webpack/blob/87660921808566ef3b8796f8df61bd79fc026108/lib/TemplatedPathPlugin.js#L19
const templateStringRegexp = /\[([\w:]+)\]/g;

export const escapeWebpackTemplateString = (s: string) =>
  s.replaceAll(templateStringRegexp, '[\\$1\\]');

function compileVanillaSource(
  loader: LoaderContext,
  externals: Externals | undefined,
): Promise<CompilationResult> {
  return new Promise((resolve, reject) => {
    const isWebpack5 = Boolean(
      loader._compiler.webpack && loader._compiler.webpack.version,
    );
    const compat = createCompat(isWebpack5);

    // Escape webpack template strings and Next.js dynamic routes in output files so they don't get replaced
    // Non-standard escape syntax, see https://webpack.js.org/configuration/output/#template-strings
    // and https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes
    const outputOptions = {
      filename: escapeWebpackTemplateString(loader.resourcePath),
    };

    // Child compiler will compile vanilla-extract files to be evaled during compilation
    const compilerName = getCompilerName(loader.resourcePath);
    const childCompiler = getRootCompilation(loader).createChildCompiler(
      compilerName,
      outputOptions,
      [],
    );

    const NodeTemplatePlugin = compat.getNodeTemplatePlugin(loader._compiler);
    const NodeTargetPlugin = compat.getNodeTargetPlugin(loader._compiler);
    const LimitChunkCountPlugin = compat.getLimitChunkCountPlugin(
      loader._compiler,
    );
    const ExternalsPlugin = compat.getExternalsPlugin(loader._compiler);

    new NodeTemplatePlugin().apply(childCompiler);
    new NodeTargetPlugin().apply(childCompiler);

    if (compat.isWebpack5) {
      const {
        EntryOptionPlugin,
        library: { EnableLibraryPlugin },
      } = loader._compiler.webpack;

      new EnableLibraryPlugin('commonjs2').apply(childCompiler);

      EntryOptionPlugin.applyEntryOption(childCompiler, loader.context, {
        child: {
          library: {
            type: 'commonjs2',
          },
          import: [loader.resourcePath],
        },
      });
    } else {
      // Webpack 4 code. Remove once support is removed
      const { LibraryTemplatePlugin, SingleEntryPlugin } = require('webpack');

      new LibraryTemplatePlugin(null, 'commonjs2').apply(childCompiler);
      new SingleEntryPlugin(loader.context, loader.resourcePath).apply(
        childCompiler,
      );
    }

    new LimitChunkCountPlugin({ maxChunks: 1 }).apply(childCompiler);
    new ExternalsPlugin('commonjs', [
      '@vanilla-extract/css',
      '@vanilla-extract/css/fileScope',
      externals,
    ]).apply(childCompiler);

    let source: string;

    if (compat.isWebpack5) {
      childCompiler.hooks.compilation.tap(compilerName, (compilation) => {
        compilation.hooks.processAssets.tap(compilerName, () => {
          source =
            compilation.assets[loader.resourcePath] &&
            (compilation.assets[loader.resourcePath].source() as string);

          // Remove all chunk assets
          compilation.chunks.forEach((chunk) => {
            chunk.files.forEach((file) => {
              compilation.deleteAsset(file);
            });
          });
        });
      });
    } else {
      childCompiler.hooks.afterCompile.tap(compilerName, (compilation) => {
        source =
          compilation.assets[loader.resourcePath] &&
          (compilation.assets[loader.resourcePath].source() as string);

        // Remove all chunk assets
        compilation.chunks.forEach((chunk) => {
          chunk.files.forEach((file) => {
            delete compilation.assets[file];
          });
        });
      });
    }

    try {
      childCompiler.runAsChild((err, _entries, compilation) => {
        if (err) {
          return reject(err);
        }

        if (!compilation) {
          return reject(
            new Error('Missing compilation in child compiler result'),
          );
        }

        if (compilation.errors.length > 0) {
          return reject(compilation.errors[0]);
        }
        if (!source) {
          return reject(new Error("Didn't get a result from child compiler"));
        }

        resolve({
          source,
          fileDependencies: Array.from(compilation.fileDependencies),
          contextDependencies: Array.from(compilation.contextDependencies),
        });
      });
    } catch (e) {
      reject(e);
    }
  });
}
