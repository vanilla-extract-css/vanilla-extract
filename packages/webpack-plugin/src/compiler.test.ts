import { escapeWebpackTemplateString } from './compiler';

describe('escapeWebpackTemplateString()', () => {
  test.each([
    '/some/path/[...slug].js',
    '/some/path/[[...slug]]/index.js',
    '/some/path[]/[slug]/[[foo]]/index.js',
  ])('%s pattern', (filePath) => {
    expect(escapeWebpackTemplateString(filePath)).toMatchSnapshot();
  });
});
