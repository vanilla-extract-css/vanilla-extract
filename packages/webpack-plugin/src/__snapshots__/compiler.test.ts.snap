// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`escapeWebpackTemplateString() /some/path/[...slug].js pattern 1`] = `"/some/path/[...slug].js"`;

exports[`escapeWebpackTemplateString() /some/path/[[...slug]]/index.js pattern 1`] = `"/some/path/[[...slug]]/index.js"`;

exports[`escapeWebpackTemplateString() /some/path[]/[slug]/[[foo]]/index.js pattern 1`] = `"/some/path[]/[\\slug\\]/[[\\foo\\]]/index.js"`;
